realm/realm-library/src/main/cpp/io_realm_internal_Collection.cpp

```diff
@@ -0,0 +1,471 @@
+/*
+ * Copyright 2016 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <jni.h>
+#include "io_realm_internal_Collection.h"
+
+#include <vector>
+
+#include <shared_realm.hpp>
+#include <results.hpp>
+
+#include "util.hpp"
+#include "java_sort_descriptor.hpp"
+#include "jni_util/java_method.hpp"
+#include "jni_util/java_global_weak_ref.hpp"
+
+using namespace realm;
+using namespace realm::jni_util;
+using namespace realm::_impl;
+
+// We need to control the life cycle of Results, weak ref of Java Collection object and the NotificationToken.
+// Wrap all three together, so when the Java Collection object gets GCed, all three of them will be invalidated.
+struct ResultsWrapper {
+    JavaGlobalWeakRef m_collection_weak_ref;
+    NotificationToken m_notification_token;
+
+    ResultsWrapper(Results&& results)
+    : m_collection_weak_ref(), m_notification_token(), m_results(results), m_snapshot() {}
+
+    ResultsWrapper(ResultsWrapper&&) = delete;
+    ResultsWrapper& operator=(ResultsWrapper&&) = delete;
+
+    ResultsWrapper(ResultsWrapper const&) = delete;
+    ResultsWrapper& operator=(ResultsWrapper const&) = delete;
+
+    ~ResultsWrapper() {}
+
+    inline Results& get_original_results()
+    {
+        return m_results;
+    }
+
+    inline Results& get_results()
+    {
+        if (m_snapshot.get_mode() == Results::Mode::Empty) {
+            return m_results;
+        } else {
+            return m_snapshot;
+        }
+    }
+
+    inline void switch_to_snapshot()
+    {
+        if (m_snapshot.get_mode() == Results::Mode::Empty) {
+            m_snapshot = m_results.snapshot();
+        }
+    }
+
+    inline void switch_to_origin()
+    {
+        if (m_snapshot.get_mode() != Results::Mode::Empty) {
+            m_snapshot = Results();
+        }
+    }
+
+    // TODO: This is for deletion related APIs on the Collection. This is not efficient at all since it moves and
+    //       and creates TableView. Expose the reference of snapshot's TableView from Results and use that instead.
+    inline void refresh_snapshot()
+    {
+        m_snapshot = m_results.snapshot();
+    }
+
+    inline bool is_detached()
+    {
+        return m_snapshot.get_mode() != Results::Mode::Empty;
+    }
+
+private:
+    Results m_results;
+    Results m_snapshot;
+};
+
+static void finalize_results(jlong ptr);
```

Comment:
for `-Wmissing-declarations`. It is important since we need it to report errors when miss define the JNI native method.